modules/packages/ObjectSerialization.chpl

```diff
@@ -0,0 +1,906 @@
+/*
+ * Copyright 2021-2023 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ The ObjectSerialization module provides an objectSerializer and
+ objectDeserializer that implement a simple binary IO format.
+ */
+@unstable("The ObjectSerialization module is unstable. The module's name, its types, and serialization format are subject to change.")
+module ObjectSerialization {
+  private use IO;
+  private use CTypes;
+
+  /*
+    A binary Serializer that implements a simple binary format.
+
+    This Serializer supports an ``endian`` field which may be configured at
+    execution time.
+
+    See :ref:`the serializers technote<ioSerializers>` for a general overview
+    of Serializers and their usage.
+  */
+  record objectSerializer {
+    /*
+      'endian' represents the endianness of the binary output produced by this
+      Serializer.
+    */
+    const endian : ioendian = ioendian.native;
+
+    @chpldoc.nodoc
+    const _structured = true;
```

Comment:
Could this be a `param` s.t. conditionals get resolved at compile time?

Reply:
While this is technically feasible, it would result in this property becoming part of the type, which isn't desirable. I had thought of just removing this undocumented field, but it felt safer (in terms of accidental bugs) to just copy-paste the old binary(De)Serializer implementation over and fix it up later when we have more time to think about object(De)Serializer.

At the moment I don't think the related conditionals are adding any meaningful overhead.

Reply:
Okay, that makes sense. I'm good with keeping it as is.